---
title: "Guide to using CAMERA"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{development}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache = TRUE
)
```

# Background

CAMERA (Cross Ancestral MEndelian RAndomization) is a method to estimate the causal effect of an exposure on an outcome using GWAS summary statistics from different populations.The CAMERA package includes three important components:

* Instrument selection for multi-ancestral Mendelian randomization (MR) analysis
* Causal inference across multi-ancestral populations using structural equation model (SEM)
* Identification of pleiotropic effects across multiple ancestral populations

Each step is documented in the documentation.
The following analyses should run within a couple of minutes, depending on internet speed and the traffic that the [IEU GWAS database](https://gwas.mrcieu.ac.uk/) servers are experiencing. 


## Installation

To install CAMERA:

```r
install.packages("remotes")
remotes::install_github("universe77/CAMERA")
```

And then load it into your R session
```{r}
library(CAMERA)
```

## Data setup

CAMERA begins by choosing an exposure and outcome hypothesis that can be tested in multi-ancestral populations. Here, we will be estimating the causal effect of body mass index (BMI) on coronary heart disease (CHD) in European ancestry (EUR) and East Asian ancestry (EAS). 

Summary statistics data can be extracted from the IEU GWAS database using the [TwoSampleMR](https://mrcieu.github.io/TwoSampleMR/) package. A list of available traits can be obtained using:
```r
traits <- TwoSampleMR::available_outcomes()
```
You can also browse the available traits here: https://gwas.mrcieu.ac.uk/

Once you obtain the study IDs for the exposure and the outcome, open R6 class environment to run CAMERA. The minimum information required for CAMERA is the following:

* Summary statistics for the exposure and the outcome
* Population information 
* Plink (ver.1.90)
* LD reference data 

Plink (ver.1.90) and LD reference data are required to identify instruments that can be used for both populations. LD reference data can be accessed from: http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz 

```{r}
# Need to specify path to plink and LD reference data here.
# The following exposures and the outcomes will be used: Exposures: BMI (EUR, EAS); Outcomes: CHD (EUR, EAS)
x <- CAMERA$new(
  		exposure_ids=c("ieu-a-2", "bbj-a-1"), 
  		outcome_ids=c("ieu-a-7", "bbj-a-109"), 
  		pops = c("EUR", "EAS"),
  		bfiles=c("/Users/yc16575/OneDrive - University of Bristol/EUR", "/Users/yc16575/OneDrive - University of Bristol/EAS"), 		
  		plink = genetics.binaRies::get_plink_binary(), 		
  		radius=50000, 
  		clump_pop="EAS"
	)
```

Make sure that the exposures/outcomes are matched across the populations. Population 1 and 2 should have the same exposure-outcome pair. Also, instrument-trait associations should be consistent between the populations (e.g. how similar SNP-BMI association in EUR to SNP-BMI assocation in EAS). You can check this as follows:

```{r}
x$check_phenotypes(ids=x$exposure_ids)
x$check_phenotypes(ids=x$outcome_ids)
```

A value closes to 1 indicates that the instrument-trait associations are consistent between the populations. If the values is too low or too high, it is recommended to use other available instruments for the traits. 


## Running CAMERA

We can now perform the analysis, which will do the following:

1. Extract instruments for the exposures
2. Check the validity of the instruments across the populations (Standardise/scale the data if necessary)
3. Extract new instruments based on LD information and fine-mapping
4. Extract instruments for the outcomes
5. Harmonise the exposure data and the outcome data
6. Perform MR using SEM


## Step by step analysis 

See the `?CAMERA` for options on the parameters for this analysis. 


### Step 1. Extract instruments for the exposures from each pouplation

This returns a set of LD clumped SNPs that are GWAS significant for BMI (raw instruments). The following function identifies SNPs that have strong associations with the exposure in each population. This is the same method as instrument extraction for multivariable MR. A data frame of the extracted instruments is stored in `x$instrument_raw`

```{r}
x$extract_instruments()
head(x$instrument_raw)
```


### Step 2. Check the validity of the instruments

It is important to ensure that the instruments for the exposure are valid across the populations. Once instruments for the exposure trait are identified for each population, we can assess specificity of the instruments. Each of the following functions estimates heterogeneity of the instruments between and calculates fraction of the instruments (obtained from the Step 1) that are replicated between the populations.

```{r}
x$instrument_heterogeneity(x$instrument_raw)
x$estimate_instrument_specificity(x$instrument_raw, alpha = "bonferroni")
```

To correct for winner's curse:

```{r}
x$estimate_instrument_specificity(x$instrument_raw, alpha = "bonferroni", winnerscurse=TRUE)
```

### Step 3. Extract new instruments based on LD information and fine-mapping

For Cross/Multi-ancestral MR it is important to ensure that genetic structures of the populations are properly considered. The following functions will identify a new set of SNPs that have the largest effect on the exposure, considering LD patterns. For the cases where the most strongly associated SNPs are not available at exactly the same position, the function searches alternative SNPs that are located near the original SNP and show the largest effect size magnitude (MaxZ method). A data frame of the extracted instruments is stored in `x$instrument_maxz`

```{r, fig.width = 7, fig.height = 6, out.width = "700px", out.height="600px", dpi=300}
# Identify genomic regions that include the raw instruments (x$instrument_raw)
suppressMessages(x$extract_instrument_regions())
# Extract instrument from identified region by searching across region for best associations
x$scan_regional_instruments()
# Plot regions and their region-selected instruments
x$plot_regional_instruments(instruments=x$instrument_maxz)
```

Check the replication fraction between the populations (instrument_maxz).
```{r}
# Estimate what fraction of the region-selected instruments we'd expect to replicate
x$instrument_heterogeneity(x$instrument_raw)
x$estimate_instrument_specificity(x$instrument_raw, alpha = "bonferroni", winnerscurse = TRUE)
```

Also, instruments can be obtained using fine mapping method. Extracted instruments are stored in `x$instrument_susie` or `x$instrument_susie`
```r
# Calculate LD matrix for the whole regions that include the raw instruments
x$regional_ld_matrices()
# Run Susie
x$susie_finemap_regions()
# Run PAINTOR
x$paintor_finemap_regions(PAINTOR="/user/work/yc16575/PAINTOR_V3.0/PAINTOR", workdir=tempdir())
```


### Step 4. Extract instruments for the outcomes

Once instruments for the exposure trait have been specified, those variants need to be extracted from the outcome trait. You can also specify p-value thresholds.

```{r}
# Get outcome data for a particular set of instruments and run SEM model 
x$make_outcome_data(exp=x$instrument_raw, p_exp=1)
head(x$instrument_outcome)
```


### Step 5. Harmonise the exposure data and the outcome data

This creates a new data frame that has the exposure data and outcome data combined (e.g. exposure-outcome for EUR and EAS).

```{r}
x$harmonised_dat(exp=x$instrument_raw, out=x$instrument_outcome)
head(x$harmonised_dat_sem)
```


### Step 6. Perform MR using genomic structural equation model (MR-SEM)

```{r}
x$perform_basic_sem()
```

You can also perform MR-SEM using the instruments from the Step 3.
```{r}
# Standardise the betas and SEs if required
x$make_outcome_data(exp=x$instrument_maxz, p_exp=1)
head(x$instrument_outcome)
x$harmonised_dat(exp=x$instrument_maxz, out=x$instrument_outcome)
head(x$harmonised_dat_sem)
x$perform_basic_sem()
```


### Standardisation / scaling
If the units for the exposures/the outcomes are different across the populations, run:

```{r}
x$standardise_data(dat=x$instrument_raw, standardise_unit=TRUE, standardise_scale=FALSE)
head(x$standardised_instrument_raw)
x$standardise_data(dat=x$instrument_outcome, standardise_unit=TRUE, standardise_scale=FALSE)
head(x$standardised_instrument_outcome)
```
Note that standardisation is not required if the effect size of SNP-trait is in the unit of SD or log odds.

For the cases where the SNP-exposure or SNP-outcome association shows substantial difference between the population, scale the beta estimates.

```{r}
x$standardise_data(dat=x$instrument_raw, standardise_unit=TRUE, standardise_scale=TRUE)
head(x$standardised_instrument_raw)
x$standardise_data(dat=x$instrument_outcome, standardise_unit=TRUE, standardise_scale=FALSE)
head(x$standardised_outcome)
```

You can check specificity of the instruments again using:
```{r}
x$instrument_heterogeneity(x$standardised_instrument_raw)
x$estimate_instrument_specificity(x$standardised_instrument_raw, alpha = "bonferroni", winnerscurse=TRUE)
```

You can also make harmonised data again after standaridisation / scaling:
```{r}
x$harmonised_dat(exp=x$standardised_instrument_raw, out=x$standardised_outcome)
head(x$harmonised_dat_sem)
```

