<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="CAMeRa">
<title>Tutorial â€¢ CAMeRa</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial">
<meta property="og:description" content="CAMeRa">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">CAMeRa</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/import-local.html">Import to CAMERA from local data</a>
    <a class="dropdown-item" href="../articles/tutorial.html">Tutorial</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/universe77/CAMERA">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Tutorial</h1>
            
      
      
      <div class="d-none name"><code>tutorial.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CAMeRa</span><span class="op">)</span></span>
<span><span class="co">#&gt; API: public: http://gwas-api.mrcieu.ac.uk/</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This software attempts to bring together various tools to improve
cross-ancestry Mendelian randomisation. It will use an example of
performing analysis of BMI on coronary heart disease across four major
ancestral groups.</p>
<p>Overview:</p>
<ul>
<li>Initialise data</li>
<li>Check phenotype scales across ancestries</li>
<li>Extract instruments</li>
<li>Evaluate instrument heterogeneity across populations</li>
<li>Extract outcome data</li>
<li>Harmonise exposure and outcome data</li>
<li>Perform analysis using raw instruments</li>
<li>Perform regional scan to obtain LD agnostic instruments</li>
<li>Re-perform analysis using regional instruments</li>
<li>Evaluate similarity of pleiotropy across ancestry</li>
<li>Evaluate similarity of instrument-exposure associations</li>
<li>Use cross-population instrument-exposure heterogeneity in MR GxE
framework to estimate pleiotropy distributions</li>
<li>Saving and loading data</li>
</ul>
</div>
<div class="section level2">
<h2 id="initialise-data">Initialise data<a class="anchor" aria-label="anchor" href="#initialise-data"></a>
</h2>
<p>CAMeRa begins by choosing an exposure and outcome hypothesis that can
be tested in multi-ancestral populations. Here, we will be estimating
the causal effect of body mass index (BMI) on coronary heart disease
(CHD) in European (EUR), East Asian (EAS), African (AFR) and South Asian
(SAS) ancestries.</p>
<p>Summary statistics data can be extracted from the IEU GWAS database
using the <a href="https://mrcieu.github.io/TwoSampleMR/" class="external-link">TwoSampleMR</a> package. A
list of available traits can be obtained using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu">TwoSampleMR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/available_outcomes.html" class="external-link">available_outcomes</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You can also browse the available traits here: <a href="https://gwas.mrcieu.ac.uk/" class="external-link uri">https://gwas.mrcieu.ac.uk/</a>. Also see other vignettes on
this site about how you can use local summary statistics instead.</p>
<p>Once you obtain the study IDs for the exposure and the outcome, open
R6 class environment to run CAMERA. The minimum information required for
CAMERA is the following:</p>
<ul>
<li>Summary statistics for the exposure and the outcome</li>
<li>Population information</li>
<li>Plink (ver.1.90)</li>
<li>LD reference data</li>
</ul>
<p>Plink (ver.1.90) and LD reference data are required to identify
instruments that can be used for both populations. LD reference data can
be accessed from: <a href="http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz" class="external-link uri">http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bfile_dir</span> <span class="op">&lt;-</span> <span class="st">"/path/to/ld_files"</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/CAMERA.html">CAMERA</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  exposure_ids<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"ukb-e-23104_CSA"</span>, </span>
<span>    <span class="st">"ukb-e-21001_AFR"</span>, </span>
<span>    <span class="st">"ukb-b-19953"</span>, </span>
<span>    <span class="st">"bbj-a-1"</span></span>
<span>  <span class="op">)</span>, </span>
<span>  outcome_ids<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"ukb-e-411_CSA"</span>, </span>
<span>    <span class="st">"ukb-e-411_AFR"</span>, </span>
<span>    <span class="st">"ieu-a-7"</span>, </span>
<span>    <span class="st">"bbj-a-109"</span></span>
<span>  <span class="op">)</span>, </span>
<span>  pops <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SAS"</span>, <span class="st">"AFR"</span>, <span class="st">"EUR"</span>, <span class="st">"EAS"</span><span class="op">)</span>,</span>
<span>  bfiles<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">bfile_dir</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SAS"</span>, <span class="st">"AFR"</span>, <span class="st">"EUR"</span>, <span class="st">"EAS"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  plink <span class="op">=</span> <span class="fu">genetics.binaRies</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/genetics.binaRies/man/get_plink_binary.html" class="external-link">get_plink_binary</a></span><span class="op">(</span><span class="op">)</span>,        </span>
<span>  radius<span class="op">=</span><span class="fl">50000</span>, </span>
<span>  clump_pop<span class="op">=</span><span class="st">"EUR"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">x</span></span>
<span><span class="co">#&gt; &lt;CAMERA&gt;</span></span>
<span><span class="co">#&gt;   Public:</span></span>
<span><span class="co">#&gt;     assign: function (...) </span></span>
<span><span class="co">#&gt;     bfiles: /path/to/ld_files/SAS /path/to/ld_files/AFR /path/to/ld_ ...</span></span>
<span><span class="co">#&gt;     check_phenotypes: function (ids = self$exposure_ids) </span></span>
<span><span class="co">#&gt;     clone: function (deep = FALSE) </span></span>
<span><span class="co">#&gt;     clump_pop: EUR</span></span>
<span><span class="co">#&gt;     cross_estimate: function (dat = self$harmonised_dat) </span></span>
<span><span class="co">#&gt;     estimate_instrument_heterogeneity_per_variant: function (dat = self$harmonised_dat) </span></span>
<span><span class="co">#&gt;     estimate_instrument_specificity: function (instrument, alpha = "bonferroni", winnerscurse = FALSE) </span></span>
<span><span class="co">#&gt;     expected_replications: NULL</span></span>
<span><span class="co">#&gt;     exposure_ids: ukb-e-23104_CSA ukb-e-21001_AFR ukb-b-19953 bbj-a-1</span></span>
<span><span class="co">#&gt;     exposure_metadata: GwasInfo, data.frame</span></span>
<span><span class="co">#&gt;     extract_instrument_regions: function (radius = self$radius, instrument_raw = self$instrument_raw, </span></span>
<span><span class="co">#&gt;     extract_instruments: function (exposure_ids = self$exposure_ids, ...) </span></span>
<span><span class="co">#&gt;     fema_regional_instruments: function (method = "fema", instrument_regions = self$instrument_regions, </span></span>
<span><span class="co">#&gt;     get_metadata: function (exposure_ids = self$exposure_ids, outcome_ids = self$outcome_ids) </span></span>
<span><span class="co">#&gt;     harmonise: function (exp = self$instrument_raw, out = self$instrument_outcome) </span></span>
<span><span class="co">#&gt;     harmonise_deprecated: function (exp = self$instrument_raw, out = self$instrument_outcome) </span></span>
<span><span class="co">#&gt;     harmonised_dat: NULL</span></span>
<span><span class="co">#&gt;     harmonised_dat_sem: NULL</span></span>
<span><span class="co">#&gt;     harmonised_data_check: NULL</span></span>
<span><span class="co">#&gt;     import: function (x) </span></span>
<span><span class="co">#&gt;     import_from_local: function (instrument_raw, instrument_outcome, instrument_regions, </span></span>
<span><span class="co">#&gt;     initialize: function (exposure_ids = NULL, outcome_ids = NULL, pops = NULL, </span></span>
<span><span class="co">#&gt;     instrument_fema: NULL</span></span>
<span><span class="co">#&gt;     instrument_fema_regions: NULL</span></span>
<span><span class="co">#&gt;     instrument_heterogeneity: function (instrument = self$instrument_raw, alpha = "bonferroni", </span></span>
<span><span class="co">#&gt;     instrument_heterogeneity_per_variant: NULL</span></span>
<span><span class="co">#&gt;     instrument_maxz: NULL</span></span>
<span><span class="co">#&gt;     instrument_mscaviar: NULL</span></span>
<span><span class="co">#&gt;     instrument_outcome: NULL</span></span>
<span><span class="co">#&gt;     instrument_outcome_regions: NULL</span></span>
<span><span class="co">#&gt;     instrument_paintor: NULL</span></span>
<span><span class="co">#&gt;     instrument_raw: NULL</span></span>
<span><span class="co">#&gt;     instrument_region_zscores: NULL</span></span>
<span><span class="co">#&gt;     instrument_regions: NULL</span></span>
<span><span class="co">#&gt;     instrument_specificity: NULL</span></span>
<span><span class="co">#&gt;     instrument_specificity_summary: NULL</span></span>
<span><span class="co">#&gt;     instrument_susie: NULL</span></span>
<span><span class="co">#&gt;     ld_matrices: NULL</span></span>
<span><span class="co">#&gt;     make_outcome_data: function (exp = self$instrument_raw, p_exp = 0.05/nrow(exp)) </span></span>
<span><span class="co">#&gt;     mrgxe: function (dat = self$harmonised_dat, variant_list = subset(self$instrument_heterogeneity_per_variant, </span></span>
<span><span class="co">#&gt;     mrgxe_plot: function (mrgxe_res = self$mrgxe_res) </span></span>
<span><span class="co">#&gt;     mrgxe_plot_variant: function (variant = self$mrgxe_res %&gt;% dplyr::filter(p.adjust(a_pval, </span></span>
<span><span class="co">#&gt;     mrgxe_res: NULL</span></span>
<span><span class="co">#&gt;     mrres: NULL</span></span>
<span><span class="co">#&gt;     MsCAVIAR_finemap_regions: function (region = self$instrument_regions, ld = self$ld_matrices, </span></span>
<span><span class="co">#&gt;     mscaviar_results: NULL</span></span>
<span><span class="co">#&gt;     outcome_ids: ukb-e-411_CSA ukb-e-411_AFR ieu-a-7 bbj-a-109</span></span>
<span><span class="co">#&gt;     outcome_metadata: GwasInfo, data.frame</span></span>
<span><span class="co">#&gt;     output: list</span></span>
<span><span class="co">#&gt;     paintor_finemap_regions: function (region = self$instrument_regions, ld = self$ld_matrices, </span></span>
<span><span class="co">#&gt;     paintor_results: NULL</span></span>
<span><span class="co">#&gt;     perform_basic_sem: function (harmonised_dat = self$harmonised_dat_sem) </span></span>
<span><span class="co">#&gt;     pleiotropy: function (harmonised_dat = self$harmonised_dat) </span></span>
<span><span class="co">#&gt;     pleiotropy_agreement: NULL</span></span>
<span><span class="co">#&gt;     pleiotropy_deprecated: function (harmonised_dat = self$harmonised_dat_sem) </span></span>
<span><span class="co">#&gt;     pleiotropy_outliers: NULL</span></span>
<span><span class="co">#&gt;     pleiotropy_Q_outliers: NULL</span></span>
<span><span class="co">#&gt;     pleiotropy_specificity_deprecated: function (harmonised_dat = self$harmonised_dat_sem, sem_result = self$sem_result, </span></span>
<span><span class="co">#&gt;     plink: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resou ...</span></span>
<span><span class="co">#&gt;     plot_cross_estimate: function (res = self$mrres, qj_alpha = 0.05) </span></span>
<span><span class="co">#&gt;     plot_pleiotropy: function (dat = self$pleiotropy_outliers) </span></span>
<span><span class="co">#&gt;     plot_pleiotropy_heterogeneity: function (dat = self$pleiotropy_Q_outliers, pthresh = 0.05) </span></span>
<span><span class="co">#&gt;     plot_regional_instruments: function (region, instrument_regions = self$instrument_regions, </span></span>
<span><span class="co">#&gt;     plot_regional_instruments_maxz: function (instrument_region_zscores = self$instrument_region_zscores, </span></span>
<span><span class="co">#&gt;     pops: SAS AFR EUR EAS</span></span>
<span><span class="co">#&gt;     radius: 50000</span></span>
<span><span class="co">#&gt;     regional_ld_matrices: function (instrument_regions = self$instrument_regions, bfiles = self$bfiles, </span></span>
<span><span class="co">#&gt;     replication_evaluation: function (instrument = self$instrument_raw, ld = self$ld_matrices) </span></span>
<span><span class="co">#&gt;     runsem: function (model, data, modname) </span></span>
<span><span class="co">#&gt;     scan_regional_instruments: function (instrument_raw = self$instrument_raw, instrument_regions = self$instrument_regions) </span></span>
<span><span class="co">#&gt;     sem_result: NULL</span></span>
<span><span class="co">#&gt;     set_summary: function () </span></span>
<span><span class="co">#&gt;     source: OpenGWAS</span></span>
<span><span class="co">#&gt;     standardise_data: function (dat = self$instrument_raw, standardise_unit = FALSE, </span></span>
<span><span class="co">#&gt;     standardised_instrument_maxz: NULL</span></span>
<span><span class="co">#&gt;     standardised_instrument_mscaviar: NULL</span></span>
<span><span class="co">#&gt;     standardised_instrument_paintor: NULL</span></span>
<span><span class="co">#&gt;     standardised_instrument_raw: NULL</span></span>
<span><span class="co">#&gt;     standardised_instrument_susie: NULL</span></span>
<span><span class="co">#&gt;     standardised_outcome: NULL</span></span>
<span><span class="co">#&gt;     summary: NULL</span></span>
<span><span class="co">#&gt;     susie_finemap_regions: function (dat = self$instrument_regions, ld = self$ld_matrices) </span></span>
<span><span class="co">#&gt;     susie_results: NULL</span></span>
<span><span class="co">#&gt;   Private:</span></span>
<span><span class="co">#&gt;     allele_frequency: function (dat = dat) </span></span>
<span><span class="co">#&gt;     bootstrap: function (wr, wr.se, ivw, ivw.se, nboot = 1000) </span></span>
<span><span class="co">#&gt;     bootstrap_diff: function (nboot, slope, slope_se, b_out, b_out_se, b_exp, b_exp_se) </span></span>
<span><span class="co">#&gt;     greedy_remove: function (r, thresh) </span></span>
<span><span class="co">#&gt;     jackknife2: function (x, theta, ...) </span></span>
<span><span class="co">#&gt;     sd_standardise: function (dat = dat) </span></span>
<span><span class="co">#&gt;     susie_overlaps: function (su1, su2)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="check-phenotype-scales-across-ancestries">Check phenotype scales across ancestries<a class="anchor" aria-label="anchor" href="#check-phenotype-scales-across-ancestries"></a>
</h2>
<p>Make sure that the exposures/outcomes are matched across the
populations. The different populations should have the same
exposure-outcome pair, with exposure and outcome traits measured in the
same way and with the same units. Also, instrument-trait associations
should be consistent between the populations (e.g.Â how similar SNP-BMI
association in EUR to SNP-BMI assocation in EAS). You can check this as
follows:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">check_phenotypes</span><span class="op">(</span>ids<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">exposure_ids</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">check_phenotypes</span><span class="op">(</span>ids<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">outcome_ids</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extract-instruments">Extract instruments<a class="anchor" aria-label="anchor" href="#extract-instruments"></a>
</h2>
<p>We can now perform the analysis, which will do the following:</p>
<ol style="list-style-type: decimal">
<li>Extract instruments for the exposures</li>
<li>Check the validity of the instruments across the populations
(Standardise/scale the data if necessary)</li>
<li>Extract new instruments based on LD information and
fine-mapping</li>
<li>Extract instruments for the outcomes</li>
<li>Harmonise the exposure data and the outcome data</li>
<li>Perform MR</li>
</ol>
<p>Note see the <code><a href="../reference/CAMERA.html">?CAMERA</a></code> for options on the parameters for
this analysis.</p>
<p>The following function identifies SNPs that have strong associations
with the exposure in each population. This is the same method as
instrument extraction for multivariable MR.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">extract_instruments</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>A data frame of the extracted instruments is stored in
<code>x$instrument_raw</code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    1495 obs. of  14 variables:</span></span>
<span><span class="co">#&gt;  $ rsid      : chr  "1:2723214_A_C" "1:6657424_A_C" "1:11207269_C_T" "1:19934900_A_G" ...</span></span>
<span><span class="co">#&gt;  $ chr       : chr  "1" "1" "1" "1" ...</span></span>
<span><span class="co">#&gt;  $ position  : int  2723214 6657424 11207269 19934900 23313353 33784146 39564930 47678458 49996959 66434743 ...</span></span>
<span><span class="co">#&gt;  $ id        : chr  "ukb-e-23104_CSA" "ukb-e-23104_CSA" "ukb-e-23104_CSA" "ukb-e-23104_CSA" ...</span></span>
<span><span class="co">#&gt;  $ beta      : num  -0.0266 -0.0235 0.0171 -0.0148 -0.0169 ...</span></span>
<span><span class="co">#&gt;  $ se        : num  0.0154 0.0154 0.0192 0.0169 0.0146 ...</span></span>
<span><span class="co">#&gt;  $ p         : num  0.0841 0.1262 0.3728 0.3823 0.2496 ...</span></span>
<span><span class="co">#&gt;  $ ea        : chr  "A" "A" "C" "A" ...</span></span>
<span><span class="co">#&gt;  $ nea       : chr  "C" "C" "T" "G" ...</span></span>
<span><span class="co">#&gt;  $ eaf       : num  0.67 0.676 0.834 0.766 0.422 ...</span></span>
<span><span class="co">#&gt;  $ units     : chr  "NA" "NA" "NA" "NA" ...</span></span>
<span><span class="co">#&gt;  $ samplesize: num  8658 8658 8658 8658 8658 ...</span></span>
<span><span class="co">#&gt;  $ method    : chr  "raw" "raw" "raw" "raw" ...</span></span>
<span><span class="co">#&gt;  $ rsido     : chr  "rs4648450" "rs3866805" "rs2791643" "rs61740466" ...</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="evaluate-instrument-heterogeneity-across-populations">Evaluate instrument heterogeneity across populations<a class="anchor" aria-label="anchor" href="#evaluate-instrument-heterogeneity-across-populations"></a>
</h2>
<p>It is important to ensure that the instruments for the exposure are
valid across the populations. Once instruments for the exposure trait
are identified for each population, we can assess specificity of the
instruments. Each of the following functions estimates heterogeneity of
the instruments between and calculates fraction of the instruments
(obtained from the Step 1) that are replicated between the
populations.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">instrument_heterogeneity</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 Ã— 9</span></span></span>
<span><span class="co">#&gt;   Reference  Replication  nsnp agreement     se      pval    I2      Q    Q_pval</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ukb-b-199â€¦ ukb-e-2310â€¦   367     0.603 0.050<span style="text-decoration: underline;">5</span> 7.52<span style="color: #949494;">e</span><span style="color: #BB0000;">- 33</span> 0.123  418.  3.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-  2</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ukb-b-199â€¦ ukb-e-2100â€¦   367     0.417 0.070<span style="text-decoration: underline;">3</span> 3.14<span style="color: #949494;">e</span><span style="color: #BB0000;">-  9</span> 0.143  428.  1.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-  2</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ukb-b-199â€¦ bbj-a-1       367     0.580 0.021<span style="text-decoration: underline;">9</span> 7.07<span style="color: #949494;">e</span><span style="color: #BB0000;">-155</span> 0.658 <span style="text-decoration: underline;">1</span>074.  8.47<span style="color: #949494;">e</span><span style="color: #BB0000;">- 71</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> bbj-a-1    ukb-e-2310â€¦    45     0.642 0.111  7.04<span style="color: #949494;">e</span><span style="color: #BB0000;">-  9</span> 0.415   77.0 1.54<span style="color: #949494;">e</span><span style="color: #BB0000;">-  3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> bbj-a-1    ukb-e-2100â€¦    45     0.415 0.152  6.49<span style="color: #949494;">e</span><span style="color: #BB0000;">-  3</span> 0.426   78.4 1.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-  3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> bbj-a-1    ukb-b-19953    45     0.860 0.051<span style="text-decoration: underline;">2</span> 2.29<span style="color: #949494;">e</span><span style="color: #BB0000;">- 63</span> 0.947  842.  4.23<span style="color: #949494;">e</span><span style="color: #BB0000;">-148</span></span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">estimate_instrument_specificity</span><span class="op">(</span>instrument<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking ukb-e-23104_CSA against ukb-e-21001_AFR</span></span>
<span><span class="co">#&gt; Checking ukb-e-23104_CSA against ukb-b-19953</span></span>
<span><span class="co">#&gt; Checking ukb-e-23104_CSA against bbj-a-1</span></span>
<span><span class="co">#&gt; Checking ukb-e-21001_AFR against ukb-e-23104_CSA</span></span>
<span><span class="co">#&gt; Checking ukb-e-21001_AFR against ukb-b-19953</span></span>
<span><span class="co">#&gt; Checking ukb-e-21001_AFR against bbj-a-1</span></span>
<span><span class="co">#&gt; Checking ukb-b-19953 against ukb-e-23104_CSA</span></span>
<span><span class="co">#&gt; Checking ukb-b-19953 against ukb-e-21001_AFR</span></span>
<span><span class="co">#&gt; Checking ukb-b-19953 against bbj-a-1</span></span>
<span><span class="co">#&gt; Checking bbj-a-1 against ukb-e-23104_CSA</span></span>
<span><span class="co">#&gt; Checking bbj-a-1 against ukb-e-21001_AFR</span></span>
<span><span class="co">#&gt; Checking bbj-a-1 against ukb-b-19953</span></span>
<span><span class="co">#&gt;          discovery     replication nsnp  metric    datum       value</span></span>
<span><span class="co">#&gt; 1  ukb-e-21001_AFR ukb-e-23104_CSA    1 P-value Expected   0.9987959</span></span>
<span><span class="co">#&gt; 2  ukb-e-21001_AFR ukb-e-23104_CSA    1 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 3  ukb-e-21001_AFR ukb-e-23104_CSA    1    Sign Expected   0.9999995</span></span>
<span><span class="co">#&gt; 4  ukb-e-21001_AFR ukb-e-23104_CSA    1    Sign Observed   1.0000000</span></span>
<span><span class="co">#&gt; 5  ukb-e-21001_AFR     ukb-b-19953    1 P-value Expected   1.0000000</span></span>
<span><span class="co">#&gt; 6  ukb-e-21001_AFR     ukb-b-19953    1 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 7  ukb-e-21001_AFR     ukb-b-19953    1    Sign Expected   0.9999995</span></span>
<span><span class="co">#&gt; 8  ukb-e-21001_AFR     ukb-b-19953    1    Sign Observed   1.0000000</span></span>
<span><span class="co">#&gt; 9  ukb-e-21001_AFR         bbj-a-1    1 P-value Expected   1.0000000</span></span>
<span><span class="co">#&gt; 10 ukb-e-21001_AFR         bbj-a-1    1 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 11 ukb-e-21001_AFR         bbj-a-1    1    Sign Expected   0.9999995</span></span>
<span><span class="co">#&gt; 12 ukb-e-21001_AFR         bbj-a-1    1    Sign Observed   1.0000000</span></span>
<span><span class="co">#&gt; 13     ukb-b-19953 ukb-e-23104_CSA  367 P-value Expected   1.7683672</span></span>
<span><span class="co">#&gt; 14     ukb-b-19953 ukb-e-23104_CSA  367 P-value Observed   2.0000000</span></span>
<span><span class="co">#&gt; 15     ukb-b-19953 ukb-e-23104_CSA  367    Sign Expected 306.2709020</span></span>
<span><span class="co">#&gt; 16     ukb-b-19953 ukb-e-23104_CSA  367    Sign Observed 261.0000000</span></span>
<span><span class="co">#&gt; 17     ukb-b-19953 ukb-e-21001_AFR  367 P-value Expected   0.4150693</span></span>
<span><span class="co">#&gt; 18     ukb-b-19953 ukb-e-21001_AFR  367 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 19     ukb-b-19953 ukb-e-21001_AFR  367    Sign Expected 277.0372286</span></span>
<span><span class="co">#&gt; 20     ukb-b-19953 ukb-e-21001_AFR  367    Sign Observed 217.0000000</span></span>
<span><span class="co">#&gt; 21     ukb-b-19953         bbj-a-1  367 P-value Expected 125.8672314</span></span>
<span><span class="co">#&gt; 22     ukb-b-19953         bbj-a-1  367 P-value Observed  42.0000000</span></span>
<span><span class="co">#&gt; 23     ukb-b-19953         bbj-a-1  367    Sign Expected 362.4660856</span></span>
<span><span class="co">#&gt; 24     ukb-b-19953         bbj-a-1  367    Sign Observed 321.0000000</span></span>
<span><span class="co">#&gt; 25         bbj-a-1 ukb-e-23104_CSA   45 P-value Expected   1.4396361</span></span>
<span><span class="co">#&gt; 26         bbj-a-1 ukb-e-23104_CSA   45 P-value Observed   2.0000000</span></span>
<span><span class="co">#&gt; 27         bbj-a-1 ukb-e-23104_CSA   45    Sign Expected  41.3635330</span></span>
<span><span class="co">#&gt; 28         bbj-a-1 ukb-e-23104_CSA   45    Sign Observed  35.0000000</span></span>
<span><span class="co">#&gt; 29         bbj-a-1 ukb-e-21001_AFR   45 P-value Expected   0.2681076</span></span>
<span><span class="co">#&gt; 30         bbj-a-1 ukb-e-21001_AFR   45 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 31         bbj-a-1 ukb-e-21001_AFR   45    Sign Expected  38.0010867</span></span>
<span><span class="co">#&gt; 32         bbj-a-1 ukb-e-21001_AFR   45    Sign Observed  27.0000000</span></span>
<span><span class="co">#&gt; 33         bbj-a-1     ukb-b-19953   45 P-value Expected  44.3523980</span></span>
<span><span class="co">#&gt; 34         bbj-a-1     ukb-b-19953   45 P-value Observed  40.0000000</span></span>
<span><span class="co">#&gt; 35         bbj-a-1     ukb-b-19953   45    Sign Expected  44.9998612</span></span>
<span><span class="co">#&gt; 36         bbj-a-1     ukb-b-19953   45    Sign Observed  45.0000000</span></span>
<span><span class="co">#&gt;           pdiff</span></span>
<span><span class="co">#&gt; 1  1.000000e+00</span></span>
<span><span class="co">#&gt; 2  1.000000e+00</span></span>
<span><span class="co">#&gt; 3  1.000000e+00</span></span>
<span><span class="co">#&gt; 4  1.000000e+00</span></span>
<span><span class="co">#&gt; 5  1.000000e+00</span></span>
<span><span class="co">#&gt; 6  1.000000e+00</span></span>
<span><span class="co">#&gt; 7  1.000000e+00</span></span>
<span><span class="co">#&gt; 8  1.000000e+00</span></span>
<span><span class="co">#&gt; 9  1.000000e+00</span></span>
<span><span class="co">#&gt; 10 1.000000e+00</span></span>
<span><span class="co">#&gt; 11 1.000000e+00</span></span>
<span><span class="co">#&gt; 12 1.000000e+00</span></span>
<span><span class="co">#&gt; 13 6.981287e-01</span></span>
<span><span class="co">#&gt; 14 6.981287e-01</span></span>
<span><span class="co">#&gt; 15 3.629548e-09</span></span>
<span><span class="co">#&gt; 16 3.629548e-09</span></span>
<span><span class="co">#&gt; 17 3.398606e-01</span></span>
<span><span class="co">#&gt; 18 3.398606e-01</span></span>
<span><span class="co">#&gt; 19 5.755367e-12</span></span>
<span><span class="co">#&gt; 20 5.755367e-12</span></span>
<span><span class="co">#&gt; 21 1.064999e-23</span></span>
<span><span class="co">#&gt; 22 1.064999e-23</span></span>
<span><span class="co">#&gt; 23 3.048804e-31</span></span>
<span><span class="co">#&gt; 24 3.048804e-31</span></span>
<span><span class="co">#&gt; 25 6.557071e-01</span></span>
<span><span class="co">#&gt; 26 6.557071e-01</span></span>
<span><span class="co">#&gt; 27 2.716399e-03</span></span>
<span><span class="co">#&gt; 28 2.716399e-03</span></span>
<span><span class="co">#&gt; 29 2.357876e-01</span></span>
<span><span class="co">#&gt; 30 2.357876e-01</span></span>
<span><span class="co">#&gt; 31 6.800783e-05</span></span>
<span><span class="co">#&gt; 32 6.800783e-05</span></span>
<span><span class="co">#&gt; 33 4.670267e-04</span></span>
<span><span class="co">#&gt; 34 4.670267e-04</span></span>
<span><span class="co">#&gt; 35 1.000000e+00</span></span>
<span><span class="co">#&gt; 36 1.000000e+00</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extract-outcome-data">Extract outcome data<a class="anchor" aria-label="anchor" href="#extract-outcome-data"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">make_outcome_data</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="harmonise-exposure-and-outcome-data">Harmonise exposure and outcome data<a class="anchor" aria-label="anchor" href="#harmonise-exposure-and-outcome-data"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">harmonise</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 Ã— 4</span></span></span>
<span><span class="co">#&gt;   pops  exposure_ids    outcome_ids   source  </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> SAS   ukb-e-23104_CSA ukb-e-411_CSA OpenGWAS</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> AFR   ukb-e-21001_AFR ukb-e-411_AFR OpenGWAS</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> EUR   ukb-b-19953     ieu-a-7       OpenGWAS</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> EAS   bbj-a-1         bbj-a-109     OpenGWAS</span></span>
<span><span class="co">#&gt; 'data.frame':    1495 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;  $ SNP : chr  "1:2723214_A_C" "1:6657424_A_C" "1:11207269_C_T" "1:19934900_A_G" ...</span></span>
<span><span class="co">#&gt;  $ pops: chr  "SAS" "SAS" "SAS" "SAS" ...</span></span>
<span><span class="co">#&gt;  $ beta: num  -0.0266 -0.0235 0.0171 -0.0148 -0.0169 ...</span></span>
<span><span class="co">#&gt;  $ se  : num  0.0154 0.0154 0.0192 0.0169 0.0146 ...</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; 'data.frame':    2434 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;  $ SNP : chr  "1:11207269_C_T" "1:11207269_C_T" "1:11207269_C_T" "1:11207269_C_T" ...</span></span>
<span><span class="co">#&gt;  $ pops: chr  "SAS" "AFR" "EUR" "EAS" ...</span></span>
<span><span class="co">#&gt;  $ beta: num  -0.06663 -0.09457 -0.00328 0.0234 0.00391 ...</span></span>
<span><span class="co">#&gt;  $ se  : num  0.0594 0.0809 0.0112 0.0319 0.0457 ...</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; 'data.frame':    1477 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;  $ SNP   : chr  "1:2723214_A_C" "1:6657424_A_C" "1:11207269_C_T" "1:19934900_A_G" ...</span></span>
<span><span class="co">#&gt;  $ pops  : chr  "SAS" "SAS" "SAS" "SAS" ...</span></span>
<span><span class="co">#&gt;  $ beta.x: num  -0.0266 -0.0235 0.0171 -0.0148 -0.0169 ...</span></span>
<span><span class="co">#&gt;  $ se.x  : num  0.0154 0.0154 0.0192 0.0169 0.0146 ...</span></span>
<span><span class="co">#&gt;  $ beta.y: num  -0.0152 -0.0295 -0.0666 0.0603 0.0465 ...</span></span>
<span><span class="co">#&gt;  $ se.y  : num  0.048 0.0479 0.0594 0.0527 0.0456 ...</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="perform-analysis-using-raw-instruments">Perform analysis using raw instruments<a class="anchor" aria-label="anchor" href="#perform-analysis-using-raw-instruments"></a>
</h2>
<p>We will perform an inverse variance weighted fixed effects MR method
within population and across all populations. Heterogeneity estimates
will be generated to evaluate if each population has a distinct
association compared to others. The combined estimate across all
populations assumes that the effect is drawn from the same distribution
and if that assumption holds then then power is improved because of the
combined information.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">cross_estimate</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; $coefficients</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 Ã— 8</span></span></span>
<span><span class="co">#&gt;   pops  Estimate `Std. Error` `t value` `Pr(&gt;|t|)`     Qj Qjpval   Qdf</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> All      0.350       0.030<span style="text-decoration: underline;">3</span>     11.5    1.52<span style="color: #949494;">e</span><span style="color: #BB0000;">-29</span> 3.14    0.371     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> AFR      0.318       0.222       1.43   1.53<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span> 0.020<span style="text-decoration: underline;">3</span>  0.887     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> EAS      0.476       0.078<span style="text-decoration: underline;">4</span>      6.07   1.62<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span> 2.60    0.107     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> EUR      0.325       0.034<span style="text-decoration: underline;">2</span>      9.52   6.61<span style="color: #949494;">e</span><span style="color: #BB0000;">-21</span> 0.503   0.478     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> SAS      0.366       0.142       2.58   9.98<span style="color: #949494;">e</span><span style="color: #BB0000;">- 3</span> 0.014<span style="text-decoration: underline;">1</span>  0.906     1</span></span></code></pre></div>
<p>Here we see a very consistent assocation across all ancestries. Note
that the <code>All</code> estimate is slightly more precisely estimated
than any of the others because it is combining similar estimates.</p>
<p>We can visualise the estimates:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">plot_cross_estimate</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="perform-regional-scan-to-obtain-ld-agnostic-instruments">Perform regional scan to obtain LD agnostic instruments<a class="anchor" aria-label="anchor" href="#perform-regional-scan-to-obtain-ld-agnostic-instruments"></a>
</h2>
<p>We can re-select the instruments scanning across the region. The
intention here is to allow all populations to contribute to a fixed
effects meta analysis to account for LD. Then the top variant in the
region from the meta analysis is used as the instrument for all
populations.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">extract_instrument_regions</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This has extracted regions around all the pooled instruments from
each exposure dataset. Now we can use either fixed effects meta analysis
within each region to choose the best SNP across all ancestries. For the
cases where the most strongly associated SNPs are not available at
exactly the same position, the function searches alternative SNPs that
are located near the original SNP and show the largest effect size
magnitude.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">fema_regional_instruments</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; tibble [1,492 Ã— 13] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;  $ beta    : num [1:1492] 0.0174 -0.0312 0.0144 0.012 0.0258 ...</span></span>
<span><span class="co">#&gt;  $ n       : num [1:1492] NA NA 461460 NA NA ...</span></span>
<span><span class="co">#&gt;  $ se      : num [1:1492] 0.01609 0.02562 0.00199 0.00422 0.01579 ...</span></span>
<span><span class="co">#&gt;  $ position: int [1:1492] 2722848 2722848 2722848 2722848 6694927 6694927 6694927 6694927 11236410 11236410 ...</span></span>
<span><span class="co">#&gt;  $ p       : num [1:1492] 2.80e-01 2.24e-01 5.10e-13 4.33e-03 1.02e-01 ...</span></span>
<span><span class="co">#&gt;  $ chr     : chr [1:1492] "1" "1" "1" "1" ...</span></span>
<span><span class="co">#&gt;  $ id      : chr [1:1492] "ukb-e-23104_CSA" "ukb-e-21001_AFR" "ukb-b-19953" "bbj-a-1" ...</span></span>
<span><span class="co">#&gt;  $ rsid    : chr [1:1492] "1:2722848_C_T" "1:2722848_C_T" "1:2722848_C_T" "1:2722848_C_T" ...</span></span>
<span><span class="co">#&gt;  $ ea      : chr [1:1492] "C" "C" "C" "C" ...</span></span>
<span><span class="co">#&gt;  $ nea     : chr [1:1492] "T" "T" "T" "T" ...</span></span>
<span><span class="co">#&gt;  $ eaf     : num [1:1492] 0.285 0.141 0.534 0.565 0.286 ...</span></span>
<span><span class="co">#&gt;  $ trait   : chr [1:1492] "Body mass index (BMI)" "Body mass index (BMI)" "Body mass index (BMI)" "Body mass index" ...</span></span>
<span><span class="co">#&gt;  $ rsido   : chr [1:1492] "rs6692145" "rs6692145" "rs6692145" "rs6692145" ...</span></span></code></pre></div>
<p>or use a Z-score based meta analysis if you are unsure about whether
the effect size scales are sufficiently consistent across the
studies:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">fema_regional_instruments</span><span class="op">(</span>method<span class="op">=</span><span class="st">"zma"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; tibble [1,492 Ã— 13] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;  $ beta    : num [1:1492] 0.0174 -0.0312 0.0144 0.012 0.027 ...</span></span>
<span><span class="co">#&gt;  $ n       : num [1:1492] NA NA 461460 NA NA ...</span></span>
<span><span class="co">#&gt;  $ se      : num [1:1492] 0.01609 0.02562 0.00199 0.00422 0.01562 ...</span></span>
<span><span class="co">#&gt;  $ position: int [1:1492] 2722848 2722848 2722848 2722848 6684906 6684906 6684906 6684906 11236410 11236410 ...</span></span>
<span><span class="co">#&gt;  $ p       : num [1:1492] 2.80e-01 2.24e-01 5.10e-13 4.33e-03 8.35e-02 ...</span></span>
<span><span class="co">#&gt;  $ chr     : chr [1:1492] "1" "1" "1" "1" ...</span></span>
<span><span class="co">#&gt;  $ id      : chr [1:1492] "ukb-e-23104_CSA" "ukb-e-21001_AFR" "ukb-b-19953" "bbj-a-1" ...</span></span>
<span><span class="co">#&gt;  $ rsid    : chr [1:1492] "1:2722848_C_T" "1:2722848_C_T" "1:2722848_C_T" "1:2722848_C_T" ...</span></span>
<span><span class="co">#&gt;  $ ea      : chr [1:1492] "C" "C" "C" "C" ...</span></span>
<span><span class="co">#&gt;  $ nea     : chr [1:1492] "T" "T" "T" "T" ...</span></span>
<span><span class="co">#&gt;  $ eaf     : num [1:1492] 0.285 0.141 0.534 0.565 0.298 ...</span></span>
<span><span class="co">#&gt;  $ trait   : chr [1:1492] "Body mass index (BMI)" "Body mass index (BMI)" "Body mass index (BMI)" "Body mass index" ...</span></span>
<span><span class="co">#&gt;  $ rsido   : chr [1:1492] "rs6692145" "rs6692145" "rs6692145" "rs6692145" ...</span></span></code></pre></div>
<p>Example</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">plot_regional_instruments</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_regions</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>Here, the European top hit is in LD with many other variants in the
European ancestry, but meta-analysing with the other ancestries
identified a similarly strongly associated variant in Europeans that is
also strongly associated in East Asians. Had we used the European top
hit then we would have missed the stronger association in the region
that associates with other ancestries too. Note that the sample sizes
for the SAS and AFR studies are very small and relatively underpowered
throughout the analysis.</p>
</div>
<div class="section level2">
<h2 id="re-perform-analysis-using-regional-instruments">Re-perform analysis using regional instruments<a class="anchor" aria-label="anchor" href="#re-perform-analysis-using-regional-instruments"></a>
</h2>
<p>Note that weâ€™re now making a copy of <code>x</code> to change the
<code>x$outcome_outcome</code> and <code>x$harmonised_dat</code> objects
to reflect the regional instrument selection. But because <code>x</code>
is an R6 class object, we need to explicitly use the
<code>x$clone()</code> function to create a new copy.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x1</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="fu">clone</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Now override the harmonised data using the regional instrument
selection</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">make_outcome_data</span><span class="op">(</span>exp<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_fema</span><span class="op">)</span></span></code></pre></div>
<p>Re-harmonise with the newly extracted data</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">harmonise</span><span class="op">(</span>exp<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_fema</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 Ã— 4</span></span></span>
<span><span class="co">#&gt;   pops  exposure_ids    outcome_ids   source  </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> SAS   ukb-e-23104_CSA ukb-e-411_CSA OpenGWAS</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> AFR   ukb-e-21001_AFR ukb-e-411_AFR OpenGWAS</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> EUR   ukb-b-19953     ieu-a-7       OpenGWAS</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> EAS   bbj-a-1         bbj-a-109     OpenGWAS</span></span>
<span><span class="co">#&gt; tibble [1,492 Ã— 4] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;  $ SNP : chr [1:1492] "1:2722848_C_T" "1:2722848_C_T" "1:2722848_C_T" "1:2722848_C_T" ...</span></span>
<span><span class="co">#&gt;  $ pops: chr [1:1492] "SAS" "AFR" "EUR" "EAS" ...</span></span>
<span><span class="co">#&gt;  $ beta: num [1:1492] 0.0174 -0.0312 0.0144 0.012 0.027 ...</span></span>
<span><span class="co">#&gt;  $ se  : num [1:1492] 0.01609 0.02562 0.00199 0.00422 0.01562 ...</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; 'data.frame':    2434 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;  $ SNP : chr  "1:11207269_C_T" "1:11207269_C_T" "1:11207269_C_T" "1:11207269_C_T" ...</span></span>
<span><span class="co">#&gt;  $ pops: chr  "SAS" "AFR" "EUR" "EAS" ...</span></span>
<span><span class="co">#&gt;  $ beta: num  -0.06663 -0.09457 -0.00328 0.0234 0.00391 ...</span></span>
<span><span class="co">#&gt;  $ se  : num  0.0594 0.0809 0.0112 0.0319 0.0457 ...</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; tibble [1,487 Ã— 6] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;  $ SNP   : chr [1:1487] "1:2722848_C_T" "1:2722848_C_T" "1:2722848_C_T" "1:2722848_C_T" ...</span></span>
<span><span class="co">#&gt;  $ pops  : chr [1:1487] "SAS" "AFR" "EUR" "EAS" ...</span></span>
<span><span class="co">#&gt;  $ beta.x: num [1:1487] 0.0174 -0.0312 0.0144 0.012 0.027 ...</span></span>
<span><span class="co">#&gt;  $ se.x  : num [1:1487] 0.01609 0.02562 0.00199 0.00422 0.01562 ...</span></span>
<span><span class="co">#&gt;  $ beta.y: num [1:1487] -0.00619 -0.07145 0.00269 -0.01034 0.05529 ...</span></span>
<span><span class="co">#&gt;  $ se.y  : num [1:1487] 0.05008 0.1119 0.00969 0.01632 0.04868 ...</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>Re-estimate the MR associations using the newly derived regional
instruments</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">cross_estimate</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; $coefficients</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 Ã— 8</span></span></span>
<span><span class="co">#&gt;   pops  Estimate `Std. Error` `t value` `Pr(&gt;|t|)`    Qj Qjpval   Qdf</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> All      0.460       0.029<span style="text-decoration: underline;">1</span>     15.8    3.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-52</span> 1.91   0.591     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> AFR      0.379       0.193       1.96   4.97<span style="color: #949494;">e</span><span style="color: #BB0000;">- 2</span> 0.176  0.675     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> EAS      0.544       0.071<span style="text-decoration: underline;">2</span>      7.65   3.69<span style="color: #949494;">e</span><span style="color: #BB0000;">-14</span> 1.38   0.240     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> EUR      0.448       0.033<span style="text-decoration: underline;">4</span>     13.4    5.99<span style="color: #949494;">e</span><span style="color: #BB0000;">-39</span> 0.131  0.718     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> SAS      0.396       0.136       2.92   3.52<span style="color: #949494;">e</span><span style="color: #BB0000;">- 3</span> 0.226  0.635     1</span></span></code></pre></div>
<p>The precision of the associations are improved because of the
improved instrument selection.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">plot_cross_estimate</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p>Evaluate instrument specificity. First using heterogeneity</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">instrument_heterogeneity</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_fema</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 Ã— 9</span></span></span>
<span><span class="co">#&gt;   Reference  Replication  nsnp agreement     se      pval     I2      Q   Q_pval</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ukb-b-199â€¦ ukb-e-2310â€¦   370     0.685 0.049<span style="text-decoration: underline;">7</span> 3.25<span style="color: #949494;">e</span><span style="color: #BB0000;">- 43</span> 0.101   412.  6.21<span style="color: #949494;">e</span><span style="color: #BB0000;">- 2</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ukb-b-199â€¦ ukb-e-2100â€¦   370     0.540 0.078<span style="text-decoration: underline;">1</span> 4.58<span style="color: #949494;">e</span><span style="color: #BB0000;">- 12</span> 0.305   533.  4.89<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ukb-b-199â€¦ bbj-a-1       370     0.689 0.022<span style="text-decoration: underline;">8</span> 2.46<span style="color: #949494;">e</span><span style="color: #BB0000;">-200</span> 0.680  <span style="text-decoration: underline;">1</span>155.  9.07<span style="color: #949494;">e</span><span style="color: #BB0000;">-82</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> bbj-a-1    ukb-e-2310â€¦    59     0.809 0.079<span style="text-decoration: underline;">0</span> 1.28<span style="color: #949494;">e</span><span style="color: #BB0000;">- 24</span> 0.027<span style="text-decoration: underline;">5</span>   60.7 3.80<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> bbj-a-1    ukb-e-2100â€¦    59     0.718 0.136  1.18<span style="color: #949494;">e</span><span style="color: #BB0000;">-  7</span> 0.311    85.7 1.05<span style="color: #949494;">e</span><span style="color: #BB0000;">- 2</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> bbj-a-1    ukb-b-19953    59     0.877 0.035<span style="text-decoration: underline;">3</span> 3.64<span style="color: #949494;">e</span><span style="color: #BB0000;">-136</span> 0.905   619.  7.37<span style="color: #949494;">e</span><span style="color: #BB0000;">-95</span></span></span></code></pre></div>
<p>Compared to above the <code>agreement</code> regression slopes are
much closer to 1 for all ancestries.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">estimate_instrument_specificity</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_fema</span>, alpha <span class="op">=</span> <span class="st">"bonferroni"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking ukb-e-23104_CSA against ukb-e-21001_AFR</span></span>
<span><span class="co">#&gt; Checking ukb-e-23104_CSA against ukb-b-19953</span></span>
<span><span class="co">#&gt; Checking ukb-e-23104_CSA against bbj-a-1</span></span>
<span><span class="co">#&gt; Checking ukb-e-21001_AFR against ukb-e-23104_CSA</span></span>
<span><span class="co">#&gt; Checking ukb-e-21001_AFR against ukb-b-19953</span></span>
<span><span class="co">#&gt; Checking ukb-e-21001_AFR against bbj-a-1</span></span>
<span><span class="co">#&gt; Checking ukb-b-19953 against ukb-e-23104_CSA</span></span>
<span><span class="co">#&gt; Checking ukb-b-19953 against ukb-e-21001_AFR</span></span>
<span><span class="co">#&gt; Checking ukb-b-19953 against bbj-a-1</span></span>
<span><span class="co">#&gt; Checking bbj-a-1 against ukb-e-23104_CSA</span></span>
<span><span class="co">#&gt; Checking bbj-a-1 against ukb-e-21001_AFR</span></span>
<span><span class="co">#&gt; Checking bbj-a-1 against ukb-b-19953</span></span>
<span><span class="co">#&gt;          discovery     replication nsnp  metric    datum       value</span></span>
<span><span class="co">#&gt; 1  ukb-e-21001_AFR ukb-e-23104_CSA    1 P-value Expected   0.9987977</span></span>
<span><span class="co">#&gt; 2  ukb-e-21001_AFR ukb-e-23104_CSA    1 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 3  ukb-e-21001_AFR ukb-e-23104_CSA    1    Sign Expected   0.9999995</span></span>
<span><span class="co">#&gt; 4  ukb-e-21001_AFR ukb-e-23104_CSA    1    Sign Observed   1.0000000</span></span>
<span><span class="co">#&gt; 5  ukb-e-21001_AFR     ukb-b-19953    1 P-value Expected   1.0000000</span></span>
<span><span class="co">#&gt; 6  ukb-e-21001_AFR     ukb-b-19953    1 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 7  ukb-e-21001_AFR     ukb-b-19953    1    Sign Expected   0.9999995</span></span>
<span><span class="co">#&gt; 8  ukb-e-21001_AFR     ukb-b-19953    1    Sign Observed   1.0000000</span></span>
<span><span class="co">#&gt; 9  ukb-e-21001_AFR         bbj-a-1    1 P-value Expected   1.0000000</span></span>
<span><span class="co">#&gt; 10 ukb-e-21001_AFR         bbj-a-1    1 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 11 ukb-e-21001_AFR         bbj-a-1    1    Sign Expected   0.9999995</span></span>
<span><span class="co">#&gt; 12 ukb-e-21001_AFR         bbj-a-1    1    Sign Observed   1.0000000</span></span>
<span><span class="co">#&gt; 13     ukb-b-19953 ukb-e-23104_CSA  370 P-value Expected   1.8157266</span></span>
<span><span class="co">#&gt; 14     ukb-b-19953 ukb-e-23104_CSA  370 P-value Observed   2.0000000</span></span>
<span><span class="co">#&gt; 15     ukb-b-19953 ukb-e-23104_CSA  370    Sign Expected 309.1513616</span></span>
<span><span class="co">#&gt; 16     ukb-b-19953 ukb-e-23104_CSA  370    Sign Observed 267.0000000</span></span>
<span><span class="co">#&gt; 17     ukb-b-19953 ukb-e-21001_AFR  370 P-value Expected   0.3249830</span></span>
<span><span class="co">#&gt; 18     ukb-b-19953 ukb-e-21001_AFR  370 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 19     ukb-b-19953 ukb-e-21001_AFR  370    Sign Expected 280.3798398</span></span>
<span><span class="co">#&gt; 20     ukb-b-19953 ukb-e-21001_AFR  370    Sign Observed 224.0000000</span></span>
<span><span class="co">#&gt; 21     ukb-b-19953         bbj-a-1  370 P-value Expected 125.4393111</span></span>
<span><span class="co">#&gt; 22     ukb-b-19953         bbj-a-1  370 P-value Observed  57.0000000</span></span>
<span><span class="co">#&gt; 23     ukb-b-19953         bbj-a-1  370    Sign Expected 365.6500239</span></span>
<span><span class="co">#&gt; 24     ukb-b-19953         bbj-a-1  370    Sign Observed 326.0000000</span></span>
<span><span class="co">#&gt; 25         bbj-a-1 ukb-e-23104_CSA   59 P-value Expected   1.5690627</span></span>
<span><span class="co">#&gt; 26         bbj-a-1 ukb-e-23104_CSA   59 P-value Observed   2.0000000</span></span>
<span><span class="co">#&gt; 27         bbj-a-1 ukb-e-23104_CSA   59    Sign Expected  53.7020322</span></span>
<span><span class="co">#&gt; 28         bbj-a-1 ukb-e-23104_CSA   59    Sign Observed  49.0000000</span></span>
<span><span class="co">#&gt; 29         bbj-a-1 ukb-e-21001_AFR   59 P-value Expected   0.2168368</span></span>
<span><span class="co">#&gt; 30         bbj-a-1 ukb-e-21001_AFR   59 P-value Observed   1.0000000</span></span>
<span><span class="co">#&gt; 31         bbj-a-1 ukb-e-21001_AFR   59    Sign Expected  49.0352256</span></span>
<span><span class="co">#&gt; 32         bbj-a-1 ukb-e-21001_AFR   59    Sign Observed  40.0000000</span></span>
<span><span class="co">#&gt; 33         bbj-a-1     ukb-b-19953   59 P-value Expected  58.5784421</span></span>
<span><span class="co">#&gt; 34         bbj-a-1     ukb-b-19953   59 P-value Observed  56.0000000</span></span>
<span><span class="co">#&gt; 35         bbj-a-1     ukb-b-19953   59    Sign Expected  58.9999029</span></span>
<span><span class="co">#&gt; 36         bbj-a-1     ukb-b-19953   59    Sign Observed  59.0000000</span></span>
<span><span class="co">#&gt;           pdiff</span></span>
<span><span class="co">#&gt; 1  1.000000e+00</span></span>
<span><span class="co">#&gt; 2  1.000000e+00</span></span>
<span><span class="co">#&gt; 3  1.000000e+00</span></span>
<span><span class="co">#&gt; 4  1.000000e+00</span></span>
<span><span class="co">#&gt; 5  1.000000e+00</span></span>
<span><span class="co">#&gt; 6  1.000000e+00</span></span>
<span><span class="co">#&gt; 7  1.000000e+00</span></span>
<span><span class="co">#&gt; 8  1.000000e+00</span></span>
<span><span class="co">#&gt; 9  1.000000e+00</span></span>
<span><span class="co">#&gt; 10 1.000000e+00</span></span>
<span><span class="co">#&gt; 11 1.000000e+00</span></span>
<span><span class="co">#&gt; 12 1.000000e+00</span></span>
<span><span class="co">#&gt; 13 7.044128e-01</span></span>
<span><span class="co">#&gt; 14 7.044128e-01</span></span>
<span><span class="co">#&gt; 15 3.380126e-08</span></span>
<span><span class="co">#&gt; 16 3.380126e-08</span></span>
<span><span class="co">#&gt; 17 2.775636e-01</span></span>
<span><span class="co">#&gt; 18 2.775636e-01</span></span>
<span><span class="co">#&gt; 19 1.129478e-10</span></span>
<span><span class="co">#&gt; 20 1.129478e-10</span></span>
<span><span class="co">#&gt; 21 1.525549e-15</span></span>
<span><span class="co">#&gt; 22 1.525549e-15</span></span>
<span><span class="co">#&gt; 23 7.530802e-30</span></span>
<span><span class="co">#&gt; 24 7.530802e-30</span></span>
<span><span class="co">#&gt; 25 6.713860e-01</span></span>
<span><span class="co">#&gt; 26 6.713860e-01</span></span>
<span><span class="co">#&gt; 27 4.016606e-02</span></span>
<span><span class="co">#&gt; 28 4.016606e-02</span></span>
<span><span class="co">#&gt; 29 1.952602e-01</span></span>
<span><span class="co">#&gt; 30 1.952602e-01</span></span>
<span><span class="co">#&gt; 31 4.477276e-03</span></span>
<span><span class="co">#&gt; 32 4.477276e-03</span></span>
<span><span class="co">#&gt; 33 8.803643e-03</span></span>
<span><span class="co">#&gt; 34 8.803643e-03</span></span>
<span><span class="co">#&gt; 35 1.000000e+00</span></span>
<span><span class="co">#&gt; 36 1.000000e+00</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="va">instrument_specificity</span><span class="op">$</span><span class="va">distinct</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">table</span></span>
<span><span class="co">#&gt; .</span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;  1124   166</span></span></code></pre></div>
<p>The number of apparently distinct instruments is reduced also.</p>
<!-- ## Attempt to re-scale instrument-exposure associations -->
</div>
<div class="section level2">
<h2 id="evaluate-similarity-of-pleiotropy-across-ancestry">Evaluate similarity of pleiotropy across ancestry<a class="anchor" aria-label="anchor" href="#evaluate-similarity-of-pleiotropy-across-ancestry"></a>
</h2>
<p>Note: the term pleiotropy used here refers to â€˜horizontal
pleiotropyâ€™, the influence of the SNP on the outcome not mediated
through the exposure.</p>
<p>Here we will find pleiotropy outliers from the MR analysis, and then
determine if the deviation of those outliers from the overall MR
estimates is consistent across populations</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">pleiotropy</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Outliers are detected based on whether a SNPâ€™s Wald ratio (in a
particular population) is substantially different from the overall
estimate. The overall estimate is the combined meta-analysis across all
populations and all SNPs, unless that populationâ€™s overall MR estimate
contributed substantially to heterogeneity. In that case, deviation is
estimated based on the populationâ€™s specific MR estimate.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="va">pleiotropy_outliers</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 36 Ã— 16</span></span></span>
<span><span class="co">#&gt;    SNP         pops    beta.x    se.x   beta.y    se.y   pval.x  pval.y       wr</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 10:1049422â€¦ SAS    0.015<span style="text-decoration: underline;">4</span>  0.017<span style="text-decoration: underline;">6</span>   3.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 0.054<span style="text-decoration: underline;">6</span>  1.90<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span> 2.91<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>  1.95<span style="color: #949494;">e</span>+0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 10:1049422â€¦ AFR    0.073<span style="text-decoration: underline;">2</span>  0.086<span style="text-decoration: underline;">7</span>  -<span style="color: #BB0000;">1.10</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> 0.385   1.99<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span> 5.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> -<span style="color: #BB0000;">1.50</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 10:1049422â€¦ EUR    0.023<span style="text-decoration: underline;">8</span>  0.003<span style="text-decoration: underline;">70</span> -<span style="color: #BB0000;">7.70</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 0.014<span style="text-decoration: underline;">4</span>  6.14<span style="color: #949494;">e</span><span style="color: #BB0000;">-11</span> 4.69<span style="color: #949494;">e</span><span style="color: #BB0000;">-8</span> -<span style="color: #BB0000;">3.24</span><span style="color: #949494;">e</span>+0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 10:1049422â€¦ EAS    0.025<span style="text-decoration: underline;">3</span>  0.004<span style="text-decoration: underline;">23</span> -<span style="color: #BB0000;">2.79</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 0.017<span style="text-decoration: underline;">7</span>  1.12<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span> 5.73<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">1.10</span><span style="color: #949494;">e</span>+0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 11:1331520â€¦ SAS    0.001<span style="text-decoration: underline;">15</span> 0.014<span style="text-decoration: underline;">5</span>  -<span style="color: #BB0000;">4.07</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 0.045<span style="text-decoration: underline;">3</span>  4.68<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span> 1.84<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> -<span style="color: #BB0000;">3.55</span><span style="color: #949494;">e</span>+1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 11:1331520â€¦ AFR   -<span style="color: #BB0000;">0.102</span>   0.038<span style="text-decoration: underline;">1</span>  -<span style="color: #BB0000;">2.18</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 0.168   3.81<span style="color: #949494;">e</span><span style="color: #BB0000;">- 3</span> 4.49<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>  2.14<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 11:1331520â€¦ EUR    0.016<span style="text-decoration: underline;">4</span>  0.002<span style="text-decoration: underline;">00</span>  4.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 0.009<span style="text-decoration: underline;">74</span> 1.15<span style="color: #949494;">e</span><span style="color: #BB0000;">-16</span> 2.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span>  2.44<span style="color: #949494;">e</span>+0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 11:1331520â€¦ EAS    0.010<span style="text-decoration: underline;">3</span>  0.004<span style="text-decoration: underline;">51</span>  8.29<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> 0.018<span style="text-decoration: underline;">9</span>  1.13<span style="color: #949494;">e</span><span style="color: #BB0000;">- 2</span> 3.30<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>  8.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 11:1169473â€¦ SAS   -<span style="color: #BB0000;">0.009</span><span style="color: #BB0000; text-decoration: underline;">40</span> 0.026<span style="text-decoration: underline;">7</span>  -<span style="color: #BB0000;">7.18</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 0.083<span style="text-decoration: underline;">0</span>  3.62<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span> 1.93<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>  7.64<span style="color: #949494;">e</span>+0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 11:1169473â€¦ AFR   -<span style="color: #BB0000;">0.026</span><span style="color: #BB0000; text-decoration: underline;">8</span>  0.032<span style="text-decoration: underline;">7</span>   1.15<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 0.144   2.06<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span> 2.12<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> -<span style="color: #BB0000;">4.28</span><span style="color: #949494;">e</span>+0</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># â„¹ 26 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># â„¹ 7 more variables: wr.se &lt;dbl&gt;, biv &lt;dbl&gt;, biv.se &lt;dbl&gt;, dif &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   dif.se &lt;dbl&gt;, Qj &lt;dbl&gt;, Qjpval &lt;dbl&gt;</span></span></span></code></pre></div>
<p>For outliers from a population, are other populations showing similar
deviation to the outlier discovery population? e.g.Â look at whether the
sign is the same for outliers discovered in Europeans:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="va">pleiotropy_agreement</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">as.data.frame</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">disc</span> <span class="op">==</span> <span class="st">"EUR"</span> <span class="op">&amp;</span> <span class="va">metric</span><span class="op">==</span><span class="st">"Sign"</span><span class="op">)</span></span>
<span><span class="co">#&gt;    disc rep nsnp metric    datum    value      pdiff</span></span>
<span><span class="co">#&gt; 15  EUR SAS    8   Sign Expected 6.142359 0.09065468</span></span>
<span><span class="co">#&gt; 16  EUR SAS    8   Sign Observed 4.000000 0.09065468</span></span>
<span><span class="co">#&gt; 19  EUR AFR    8   Sign Expected 5.737231 0.04563731</span></span>
<span><span class="co">#&gt; 20  EUR AFR    8   Sign Observed 3.000000 0.04563731</span></span>
<span><span class="co">#&gt; 23  EUR EAS    8   Sign Expected 7.397345 0.01794272</span></span>
<span><span class="co">#&gt; 24  EUR EAS    8   Sign Observed 5.000000 0.01794272</span></span></code></pre></div>
<p>Look at the overall relationship of outlier deviations across
populations</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">plot_pleiotropy</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(pop1)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(pop2, SNP)`</span></span>
<span><span class="co">#&gt; `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<p>Identify any variants that showed substantial differences in
pleiotropy deviations across populations. Note that sometimes the
pleiotropy deviation estimate is unstable due to the SNP-exposure
association being very small. Unstable estimates are attempted to be
removed automatically from the heterogeneity analysis</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">plot_pleiotropy_heterogeneity</span><span class="op">(</span>pthresh<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
<p>No SNPs are showing substantial differences in pleiotropy deviation
across populations. Plot everything by relaxing the threshold</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">plot_pleiotropy_heterogeneity</span><span class="op">(</span>pthresh<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="mr-gxe">MR GxE<a class="anchor" aria-label="anchor" href="#mr-gxe"></a>
</h2>
<p>The MR GxE model aims to estimate the horizontal pleiotropic effect
of a SNP. This is achieved by estimating its effect on the outcome in a
subset of the data where the SNP is not expected to have an association
(the zero relevance group). The cross-ancestry MR analysis can attempt
to make use of this approach by identifying variants that exhibit
heterogeneity in the instrument-exposure association across ancestries.
Those instruments can then be used to estimate the pleiotropic
association by evaluating their effect on the outcome across populations
showing differential SNP-exposure associations.</p>
<p>First identify examples of heterogeneity amongst instrument-exposure
associations</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">estimate_instrument_heterogeneity_per_variant</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 373 Ã— 5</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   SNP [373]</span></span></span>
<span><span class="co">#&gt;    SNP                Qdf      Q    Qpval     Qfdr</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 10:104942244_G_T     3  0.650 0.885    0.885   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 10:118650996_C_T     3 17.7   0.000<span style="text-decoration: underline;">502</span> 0.000<span style="text-decoration: underline;">502</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 10:134007008_A_C     3  8.12  0.043<span style="text-decoration: underline;">6</span>   0.043<span style="text-decoration: underline;">6</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 10:16750129_G_T      3  0.510 0.917    0.917   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 10:18573654_A_G      3  6.63  0.084<span style="text-decoration: underline;">7</span>   0.084<span style="text-decoration: underline;">7</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 10:21830104_A_G      3  3.19  0.363    0.363   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 10:33955430_C_T      3 13.5   0.003<span style="text-decoration: underline;">75</span>  0.003<span style="text-decoration: underline;">75</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 10:53673286_A_G      3  4.26  0.235    0.235   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 10:61842645_C_T      3  6.16  0.104    0.104   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 10:65191645_G_T      3  0.621 0.892    0.892   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># â„¹ 363 more rows</span></span></span>
<span><span class="va">x</span><span class="op">$</span><span class="va">instrument_heterogeneity_per_variant</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Qfdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 95 Ã— 5</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   SNP [95]</span></span></span>
<span><span class="co">#&gt;    SNP                Qdf     Q      Qpval       Qfdr</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 10:118650996_C_T     3 17.7  0.000<span style="text-decoration: underline;">502</span>   0.000<span style="text-decoration: underline;">502</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 10:134007008_A_C     3  8.12 0.043<span style="text-decoration: underline;">6</span>     0.043<span style="text-decoration: underline;">6</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 10:33955430_C_T      3 13.5  0.003<span style="text-decoration: underline;">75</span>    0.003<span style="text-decoration: underline;">75</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 10:87490850_A_G      3  8.51 0.036<span style="text-decoration: underline;">5</span>     0.036<span style="text-decoration: underline;">5</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 10:99772885_A_G      3 18.8  0.000<span style="text-decoration: underline;">298</span>   0.000<span style="text-decoration: underline;">298</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 11:130795698_G_T     3 14.5  0.002<span style="text-decoration: underline;">32</span>    0.002<span style="text-decoration: underline;">32</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 11:13315205_C_T      3 11.9  0.007<span style="text-decoration: underline;">86</span>    0.007<span style="text-decoration: underline;">86</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 11:2858440_A_G       3 29.5  0.000<span style="text-decoration: underline;">001</span>79 0.000<span style="text-decoration: underline;">001</span>79</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 11:43648368_G_T      3 12.4  0.006<span style="text-decoration: underline;">27</span>    0.006<span style="text-decoration: underline;">27</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 11:45420233_A_G      3  8.59 0.035<span style="text-decoration: underline;">2</span>     0.035<span style="text-decoration: underline;">2</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># â„¹ 85 more rows</span></span></span></code></pre></div>
<p>Next perform MR GxE (may take a couple of minutes while bootstraping
standard errors)</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">mrgxe</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 95 Ã— 9</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   SNP [95]</span></span></span>
<span><span class="co">#&gt;    SNP                      a       b   a_se  b_se a_pval b_pval   a_mean b_mean</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 10:118650996_C_T   4.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  2.14   0.025<span style="text-decoration: underline;">7</span>  1.99 0.039<span style="text-decoration: underline;">8</span> 0.140   5.11<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  2.12 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 10:134007008_A_C   3.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  0.805  0.027<span style="text-decoration: underline;">2</span>  2.13 0.452  0.353   1.98<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  0.482</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 10:33955430_C_T    6.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> -<span style="color: #BB0000;">0.195</span>  0.037<span style="text-decoration: underline;">4</span>  1.70 0.435  0.454   1.04<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">0.349</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 10:87490850_A_G   -<span style="color: #BB0000;">1.94</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">0.030</span><span style="color: #BB0000; text-decoration: underline;">5</span> 0.043<span style="text-decoration: underline;">7</span>  1.82 0.329  0.493  -<span style="color: #BB0000;">2.79</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">0.218</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 10:99772885_A_G   -<span style="color: #BB0000;">1.95</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">1.48</span>   0.026<span style="text-decoration: underline;">8</span>  1.77 0.233  0.202  -<span style="color: #BB0000;">2.39</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">1.45</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 11:130795698_G_T  -<span style="color: #BB0000;">1.10</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  2.17   0.023<span style="text-decoration: underline;">2</span>  1.46 0.318  0.068<span style="text-decoration: underline;">6</span> -<span style="color: #BB0000;">9.49</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  1.96 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 11:13315205_C_T    1.68<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  0.283  0.026<span style="text-decoration: underline;">1</span>  1.89 0.474  0.441   2.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>  0.364</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 11:2858440_A_G    -<span style="color: #BB0000;">8.48</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  1.42   0.114   3.08 0.229  0.322  -<span style="color: #BB0000;">7.73</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  0.845</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 11:43648368_G_T    1.64<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  1.53   0.030<span style="text-decoration: underline;">9</span>  1.74 0.298  0.190   1.12<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  0.965</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 11:45420233_A_G    1.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> -<span style="color: #BB0000;">0.248</span>  0.037<span style="text-decoration: underline;">4</span>  2.43 0.500  0.459  -<span style="color: #BB0000;">1.08</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  0.406</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># â„¹ 85 more rows</span></span></span>
<span><span class="va">x</span><span class="op">$</span><span class="va">mrgxe_res</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 95 Ã— 9</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   SNP [95]</span></span></span>
<span><span class="co">#&gt;    SNP                      a       b   a_se  b_se a_pval b_pval   a_mean b_mean</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 10:118650996_C_T   4.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  2.14   0.025<span style="text-decoration: underline;">7</span>  1.99 0.039<span style="text-decoration: underline;">8</span> 0.140   5.11<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  2.12 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 10:134007008_A_C   3.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  0.805  0.027<span style="text-decoration: underline;">2</span>  2.13 0.452  0.353   1.98<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  0.482</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 10:33955430_C_T    6.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> -<span style="color: #BB0000;">0.195</span>  0.037<span style="text-decoration: underline;">4</span>  1.70 0.435  0.454   1.04<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">0.349</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 10:87490850_A_G   -<span style="color: #BB0000;">1.94</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">0.030</span><span style="color: #BB0000; text-decoration: underline;">5</span> 0.043<span style="text-decoration: underline;">7</span>  1.82 0.329  0.493  -<span style="color: #BB0000;">2.79</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">0.218</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 10:99772885_A_G   -<span style="color: #BB0000;">1.95</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">1.48</span>   0.026<span style="text-decoration: underline;">8</span>  1.77 0.233  0.202  -<span style="color: #BB0000;">2.39</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> -<span style="color: #BB0000;">1.45</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 11:130795698_G_T  -<span style="color: #BB0000;">1.10</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  2.17   0.023<span style="text-decoration: underline;">2</span>  1.46 0.318  0.068<span style="text-decoration: underline;">6</span> -<span style="color: #BB0000;">9.49</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  1.96 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 11:13315205_C_T    1.68<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  0.283  0.026<span style="text-decoration: underline;">1</span>  1.89 0.474  0.441   2.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>  0.364</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 11:2858440_A_G    -<span style="color: #BB0000;">8.48</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  1.42   0.114   3.08 0.229  0.322  -<span style="color: #BB0000;">7.73</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  0.845</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 11:43648368_G_T    1.64<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  1.53   0.030<span style="text-decoration: underline;">9</span>  1.74 0.298  0.190   1.12<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  0.965</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 11:45420233_A_G    1.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> -<span style="color: #BB0000;">0.248</span>  0.037<span style="text-decoration: underline;">4</span>  2.43 0.500  0.459  -<span style="color: #BB0000;">1.08</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  0.406</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># â„¹ 85 more rows</span></span></span></code></pre></div>
<p>This is the distribution of the estimate of the pleiotropic effect of
each SNP that showed heterogeneity</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">mrgxe_plot</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<p>Any evidence of SNPs with substantial heterogeneity?</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="va">mrgxe_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">a_pval</span>, <span class="st">"fdr"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 Ã— 9</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   SNP [6]</span></span></span>
<span><span class="co">#&gt;   SNP                    a       b   a_se  b_se  a_pval b_pval  a_mean b_mean</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 10:118650996_C_T  0.045<span style="text-decoration: underline;">0</span>  2.14   0.025<span style="text-decoration: underline;">7</span>  1.99 0.039<span style="text-decoration: underline;">8</span>  0.140   0.051<span style="text-decoration: underline;">1</span>  2.12 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 13:58259492_A_C  -<span style="color: #BB0000;">0.080</span><span style="color: #BB0000; text-decoration: underline;">4</span> -<span style="color: #BB0000;">4.47</span>   0.041<span style="text-decoration: underline;">6</span>  3.16 0.026<span style="text-decoration: underline;">6</span>  0.078<span style="text-decoration: underline;">2</span> -<span style="color: #BB0000;">0.073</span><span style="color: #BB0000; text-decoration: underline;">7</span> -<span style="color: #BB0000;">2.60</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 16:76895693_A_G   0.056<span style="text-decoration: underline;">7</span>  2.19   0.028<span style="text-decoration: underline;">8</span>  2.47 0.024<span style="text-decoration: underline;">7</span>  0.188   0.055<span style="text-decoration: underline;">5</span>  1.89 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 19:33953354_G_T  -<span style="color: #BB0000;">0.068</span><span style="color: #BB0000; text-decoration: underline;">1</span> -<span style="color: #BB0000;">0.086</span><span style="color: #BB0000; text-decoration: underline;">9</span> 0.024<span style="text-decoration: underline;">1</span>  1.81 0.002<span style="text-decoration: underline;">33</span> 0.481  -<span style="color: #BB0000;">0.063</span><span style="color: #BB0000; text-decoration: underline;">5</span>  0.121</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 3:48118703_C_T    0.040<span style="text-decoration: underline;">6</span>  1.42   0.022<span style="text-decoration: underline;">8</span>  2.65 0.037<span style="text-decoration: underline;">3</span>  0.297   0.036<span style="text-decoration: underline;">8</span>  0.630</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 9:28418511_A_G    0.083<span style="text-decoration: underline;">5</span> -<span style="color: #BB0000;">3.03</span>   0.039<span style="text-decoration: underline;">7</span>  1.84 0.017<span style="text-decoration: underline;">8</span>  0.049<span style="text-decoration: underline;">8</span>  0.081<span style="text-decoration: underline;">7</span> -<span style="color: #BB0000;">2.90</span></span></span></code></pre></div>
<p>Itâ€™s worth always checking if these look credible e.g.Â this plots the
SNP-exposure against SNP-outcome associations for the identified SNPs.
Youâ€™d expect to see a slope reflecting the causal effect estimate with
the intercept reflecting the pleiotropic association.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">mrgxe_plot_variant</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<p>In this case the associations are very noisy, and it would be
difficult to justify that they show convincing evidence of the
pleiotropy estimate.</p>
</div>
<div class="section level2">
<h2 id="saving-and-loading-data">Saving and loading data<a class="anchor" aria-label="anchor" href="#saving-and-loading-data"></a>
</h2>
<p>It can be useful to import data from one CAMERA object to another,
because for example you may have all your data, but the CAMERA class has
been updated, and you want to initialise a new class and import all the
old data into the new class.</p>
<p>Save CAMERA objects as RDS files e.g.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">x</span>, file<span class="op">=</span><span class="st">"example-CAMERA.rds"</span><span class="op">)</span></span></code></pre></div>
<p>And load them like this:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"example-CAMERA.rds"</span><span class="op">)</span></span></code></pre></div>
<p>You can import data from one CAMERA object to another like this:</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yoonsu Cho, Gibran Hemani.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
